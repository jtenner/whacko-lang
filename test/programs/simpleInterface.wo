interface Test {
  test(): i32;
}

// I'll do it
class A implements Test {
  constructor() {}

  test(): i32 {
    return i32(42);
  }
}

class B implements Test {
  constructor() {}

  test(): i32 {
    return i32(69);
  }
}

class C implements Test {
  constructor() {}

  test(): i32 {
    return i32(123);
  }
}

class D implements Test {
  constructor() {}

  test(): i32 {
    return i32(456);
  }
}

export fn callTest(myTest: Test): i32 {
  return myTest.test();
}

export fn _start(): void {
  let a = new A();
  let b = new B();
  let c = new C();
  let d = new D();

  assert(callTest(a) == i32(42));
  assert(callTest(b) == i32(69));
  assert(callTest(c) == i32(123));
  assert(callTest(d) == i32(456));
}
