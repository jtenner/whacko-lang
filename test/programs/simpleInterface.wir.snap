program:
  callable:[std/builtins.wo]gc.initialize;functionType:()=>voidType
(extern function)
  callable:[std/builtins.wo]gc.alloc;functionType:(u32Type,usizeType)=>ptrType
(extern function)
  callable:[std/builtins.wo]gc.store_field_ptr;functionType:(i32Type,ptrType,ptrType,ptrType)=>voidType
(extern function)
  callable:[test/programs/simpleInterface.wo]_start;functionType:()=>voidType
    block: entry~85
      89~inst = new: classType:[test/programs/simpleInterface.wo]A<>
      90~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]A<> 89~inst)
      91~inst = call: callable:[test/programs/simpleInterface.wo]A.constructor;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>voidType with (runtime classType:[test/programs/simpleInterface.wo]A<> 89~inst)
      92~inst = store: (variable a) = (runtime classType:[test/programs/simpleInterface.wo]A<> 89~inst)
      95~inst = new: classType:[test/programs/simpleInterface.wo]B<>
      96~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]B<> 95~inst)
      97~inst = call: callable:[test/programs/simpleInterface.wo]B.constructor;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>voidType with (runtime classType:[test/programs/simpleInterface.wo]B<> 95~inst)
      98~inst = store: (variable b) = (runtime classType:[test/programs/simpleInterface.wo]B<> 95~inst)
      101~inst = new: classType:[test/programs/simpleInterface.wo]C<>
      102~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]C<> 101~inst)
      103~inst = call: callable:[test/programs/simpleInterface.wo]C.constructor;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>voidType with (runtime classType:[test/programs/simpleInterface.wo]C<> 101~inst)
      104~inst = store: (variable c) = (runtime classType:[test/programs/simpleInterface.wo]C<> 101~inst)
      107~inst = new: classType:[test/programs/simpleInterface.wo]D<>
      108~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]D<> 107~inst)
      109~inst = call: callable:[test/programs/simpleInterface.wo]D.constructor;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>voidType with (runtime classType:[test/programs/simpleInterface.wo]D<> 107~inst)
      110~inst = store: (variable d) = (runtime classType:[test/programs/simpleInterface.wo]D<> 107~inst)
      111~inst = load: (variable a)
      112~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]A<> 111~inst)
      116~inst = call: callable:[test/programs/simpleInterface.wo]callTest;functionType:(interfaceType:[test/programs/simpleInterface.wo]Test<>)=>i32Type with (runtime classType:[test/programs/simpleInterface.wo]A<> 111~inst)
      121~inst = (i32 42)
      122~inst = 116~inst == 121~inst
      br to next~125 if (runtime i32Type 122~inst); else to falsy~124
    block: falsy~124
      128~inst = unreachable
    block: next~125
      129~inst = load: (variable b)
      130~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]B<> 129~inst)
      133~inst = call: callable:[test/programs/simpleInterface.wo]callTest;functionType:(interfaceType:[test/programs/simpleInterface.wo]Test<>)=>i32Type with (runtime classType:[test/programs/simpleInterface.wo]B<> 129~inst)
      138~inst = (i32 69)
      139~inst = 133~inst == 138~inst
      br to next~142 if (runtime i32Type 139~inst); else to falsy~141
    block: falsy~141
      145~inst = unreachable
    block: next~142
      146~inst = load: (variable c)
      147~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]C<> 146~inst)
      150~inst = call: callable:[test/programs/simpleInterface.wo]callTest;functionType:(interfaceType:[test/programs/simpleInterface.wo]Test<>)=>i32Type with (runtime classType:[test/programs/simpleInterface.wo]C<> 146~inst)
      155~inst = (i32 123)
      156~inst = 150~inst == 155~inst
      br to next~159 if (runtime i32Type 156~inst); else to falsy~158
    block: falsy~158
      162~inst = unreachable
    block: next~159
      163~inst = load: (variable d)
      164~inst = store: (variable compiler-generated) = (runtime classType:[test/programs/simpleInterface.wo]D<> 163~inst)
      167~inst = call: callable:[test/programs/simpleInterface.wo]callTest;functionType:(interfaceType:[test/programs/simpleInterface.wo]Test<>)=>i32Type with (runtime classType:[test/programs/simpleInterface.wo]D<> 163~inst)
      172~inst = (i32 456)
      173~inst = 167~inst == 172~inst
      br to next~176 if (runtime i32Type 173~inst); else to falsy~175
    block: falsy~175
      179~inst = unreachable
    block: next~176
      return (void)

  callable:[test/programs/simpleInterface.wo]A.constructor;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>voidType
    block: entry~194
      return (void)

  callable:[test/programs/simpleInterface.wo]B.constructor;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>voidType
    block: entry~192
      return (void)

  callable:[test/programs/simpleInterface.wo]C.constructor;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>voidType
    block: entry~190
      return (void)

  callable:[test/programs/simpleInterface.wo]D.constructor;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>voidType
    block: entry~188
      return (void)

  callable:[test/programs/simpleInterface.wo]callTest;functionType:(interfaceType:[test/programs/simpleInterface.wo]Test<>)=>i32Type
    block: entry~181
      182~inst = load: (variable myTest)
      186~inst = call: callable:[test/programs/simpleInterface.wo]Test.test;methodType:interfaceType:[test/programs/simpleInterface.wo]Test<>#()=>i32Type with (runtime interfaceType:[test/programs/simpleInterface.wo]Test<> 182~inst)
      return (runtime i32Type 186~inst)

  callable:[test/programs/simpleInterface.wo]Test.test;methodType:interfaceType:[test/programs/simpleInterface.wo]Test<>#()=>i32Type
(interface method)
  callable:[test/programs/simpleInterface.wo]A.test;methodType:classType:[test/programs/simpleInterface.wo]A<>#()=>i32Type
    block: entry~226
      return (i32 42)

  callable:[test/programs/simpleInterface.wo]B.test;methodType:classType:[test/programs/simpleInterface.wo]B<>#()=>i32Type
    block: entry~220
      return (i32 69)

  callable:[test/programs/simpleInterface.wo]C.test;methodType:classType:[test/programs/simpleInterface.wo]C<>#()=>i32Type
    block: entry~214
      return (i32 123)

  callable:[test/programs/simpleInterface.wo]D.test;methodType:classType:[test/programs/simpleInterface.wo]D<>#()=>i32Type
    block: entry~208
      return (i32 456)

